File: js/reddit_GPT.js
const { OpenAI } = require('openai');

async function fetchPostsFromSubreddit(subredditName) {
  const url = `https://www.reddit.com/r/${subredditName}/new.json?limit=100`;
  const options = {
    headers: {
      'User-Agent': 'My Reddit Scraper'
    }
  };

  return new Promise((resolve, reject) => {
    const req = https.get(url, options, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        if (res.statusCode !== 200) {
          reject(new Error(`Error retrieving posts: ${res.statusCode} ${res.statusMessage}`));
          return;
        }

        const response = JSON.parse(data);
        const posts = response.data.children;

        const postList = posts.map(post => ({
          title: post.data.title,
          author: post.data.author
        }));

        resolve(postList);
      });
    });

    req.on('error', (error) => {
      reject(new Error(`Error retrieving posts: ${error.message}`));
    });

    req.end();
  });
}

function analyzePostList(posts) {
  // Analyze the post list and return the results
  // ...

  // For demonstration purposes, we'll return a dummy analysis
  return {
    totalPosts: posts.length,
    exampleAnalysis: 'This is an example analysis result.'
  };
}

async function analyzePosts(posts, apiKey) {
  const openai = new OpenAI(apiKey);

  const prompt = posts.map(post => post.title).join('\n');
  const options = {
    engine: 'davinci',
    prompt,
    maxTokens: 100
  };

  try {
    const response = await openai.complete(options);
    const analysisResult = response.choices[0].text.trim();
    return analysisResult;
  } catch (error) {
    throw new Error(`Error analyzing posts: ${error.message}`);
  }
}

// Usage example
const subredditName = 'bitcoin'; // Replace with the subreddit you want to fetch posts from
const apiKey = 'YOUR_OPENAI_API_KEY'; // Add your actual OpenAI API key here

fetchPostsFromSubreddit(subredditName)
  .then(posts => {
    return analyzePosts(posts, apiKey);
  })
  .then(analysisResult => {
    console.log('Analysis result:', analysisResult);
  })
  .catch(error => {
    console.error(error.message);
  });

File: Python/reddit_gpt.py
import praw
import openai

# Reddit API credentials
reddit_client_id = '8IvnbB8h3jYOA6XUVhlz2g'
reddit_client_secret = 'zMch1-6NR1nIT_PCgGJOWEFu5_23_w'
reddit_user_agent = 'myapp'

# OpenAI API credentials
openai_api_key = 'Your_API_Key'

# Initialize Reddit API
reddit = praw.Reddit(client_id=reddit_client_id,
                     client_secret=reddit_client_secret,
                     user_agent=reddit_user_agent)

# Set up OpenAI API
openai.api_key = openai_api_key

# Subreddit to analyze
subreddit_name = 'bitcoin'

# Number of posts to fetch
num_posts = 100

# Fetch posts from the subreddit
subreddit = reddit.subreddit(subreddit_name)
posts = subreddit.new(limit=num_posts)

# Extract titles and store them in a list
post_titles = [post.title for post in posts]


# Convert list of titles to a single string
input_text = '\n'.join(post_titles)

# Call the OpenAI API for summarization
response = openai.Completion.create(
  engine='text-davinci-003',
  prompt=input_text,
  max_tokens=100,
  temperature=0.5,
  n=1,
  stop=None,
  top_p=1,
  frequency_penalty=0,
  presence_penalty=0
)

# Extract the generated summary
summary = response.choices[0].text.strip()

# Print the summary
print("Summary of trends in", subreddit_name)
print(summary)
print(post_titles)
